### PR TITLE
WIP: Finish exam partial credit

### PR DESCRIPTION
**Pull request by tbretl**
This needs three things before being done:
* [x] Handle external grading jobs (e.g., write `v2_score` in `grading_jobs_process_external.sql` if necessary).
* [x] Refactor `examQuestionAvailablePoints.ejs` and `examQuestionAwardedPoints.ejs` to put the defined function `getStringFromFloat`somewhere more appropriate.
* [x] Add/modify test suite (without modification, build will definitely fail).

### CODE REVIEW

**Comment by tbretl** [comment]
OK, frankly astonished that all checks were passed. ;-)

**Comment by tbretl** [comment]
Note that we're not providing any information to the students yet about how points are computed (e.g., from Available Points if score is not 100%).

File: sprocs/instance_questions_points_exam.sql [hunk 1]
```diff
@@ -23,31 +23,40 @@ BEGIN
...
         current_value := NULL;
         points_list := array[]::double precision[];
     ELSE
         IF cardinality(iq.points_list) > 1 THEN
             open := TRUE;
             status := 'incorrect';
             current_value := iq.points_list[1];
-            points_list := iq.points_list[2:cardinality(iq.points_list)];
+            points_list := array[]::double precision[];
+            FOR i in 1..(cardinality(iq.points_list_original)-(coalesce(iq.number_attempts, 0)+1)) LOOP
```
Comment (mwest1066) [inline sprocs/instance_questions_points_exam.sql]: Don't need `coalesce` on `iq.number_attempts` (also above). We should add a migration to do:
```sql
ALTER TABLE instance_questions ALTER COLUMN number_attempts SET NOT NULL;
```

File: sprocs/instance_questions_points_exam.sql [hunk 2]
```diff
@@ -23,31 +23,40 @@ BEGIN
     SELECT * INTO aq FROM assessment_questions WHERE id = iq.assessment_question_id;
 
     max_points := COALESCE(aq.max_points, 0);
-    highest_submission_score := greatest(submission_score, coalesce(highest_submission_score, 0));
 
-    correct := (submission_score >= 0.5);
+    -- Update points (instance_question will be closed when number_attempts exceeds bound,
+    -- so we don't have to worry about accessing a non-existent entry in points_list_original,
+    -- but use coalesce just to be safe)
+    points := coalesce(iq.points, 0) + (coalesce(iq.points_list_original[coalesce(iq.number_attempts, 0)+1], 0) * GREATEST(0, submission_score - coalesce(iq.highest_submission_score, 0)));
```
Comment (mwest1066) [inline sprocs/instance_questions_points_exam.sql]: I think we should add a special case after this points computation like:
```sql
IF (submission_score >= 1) AND (iq.points_list_original[iq.number_attempts + 1] = max_points) THEN
    points := max_points;
END IF
```

This is for the case when `points_list = [10, 10, 10, 10]` and the student gets scores `20%, 40%, 80%, 100%`. We want this to give exactly 10 points, but we could have floating point issues from computing it incrementally.

File: sprocs/instance_questions_points_exam.sql [hunk 3]
```diff
@@ -23,31 +23,40 @@ BEGIN
     SELECT * INTO aq FROM assessment_questions WHERE id = iq.assessment_question_id;
 
     max_points := COALESCE(aq.max_points, 0);
-    highest_submission_score := greatest(submission_score, coalesce(highest_submission_score, 0));
 
-    correct := (submission_score >= 0.5);
+    -- Update points (instance_question will be closed when number_attempts exceeds bound,
+    -- so we don't have to worry about accessing a non-existent entry in points_list_original,
+    -- but use coalesce just to be safe)
```
Comment (mwest1066) [inline sprocs/instance_questions_points_exam.sql]: On the line below we don't need `coalesce` around `iq.points` or `iq.points_list_original`. We are assuming that these are non-NULL.

In fact, I think we are assuming everything is non-NULL except for `iq.highest_submission_score`, which starts as NULL when `number_attempts = 0` and is non-NULL thereafter. So I think we only need a `coalesce` in the case of `coalesce(iq.highest_submission_score, 0)`.

File: sprocs/instance_questions_points_exam.sql (hunk 2 above)
Comment (tbretl) [reply sprocs/instance_questions_points_exam.sql]: I understand. (Note - I don't think check anyway that `points_list` is non-increasing. If it isn't, results are going to be strange, here and elsewhere.)

Comment (mwest1066) [reply sprocs/instance_questions_points_exam.sql]: I just filed #773 for this.

File: sprocs/instance_questions_points_exam.sql (hunk 3 above)
Comment (tbretl) [reply sprocs/instance_questions_points_exam.sql]: done

File: sprocs/instance_questions_points_exam.sql (hunk 1 above)
Comment (tbretl) [reply sprocs/instance_questions_points_exam.sql]: done

**Comment by mwest1066** [comment]
Superseded by #779

**Pull request closed without merging**